src/node_sets.cpp

```diff
@@ -0,0 +1,320 @@
+        } else if (el.value().is_array()) {
+            const auto array = el.value();
+
+            if (array[0].is_number()) {
+                if (attribute == "population") {
+                    throw SonataError("'population' must be a string");
+                }
+
+                std::vector<int64_t> values;
+                for (auto& inner_el : array.items()) {
+                    values.emplace_back(get_integer_or_throw(inner_el.value()));
+                }
+
+                if (attribute == "node_id") {
+                    Selection::Values node_ids;
+                    std::transform(begin(values),
+                                   end(values),
+                                   back_inserter(node_ids),
+                                   [](int64_t integer) {
+                                       if (integer < 0) {
+                                           throw SonataError("'node_id' must be positive");
+                                       }
+                                       return static_cast<Selection::Value>(integer);
+                                   });
+                    ret.emplace_back(new NodeSetBasicNodeIds(std::move(node_ids)));
+                } else {
+                    ret.emplace_back(new NodeSetBasicRule<int64_t>(attribute, std::move(values)));
+                }
+            } else if (array[0].is_string()) {
+                if (attribute == "node_id") {
+                    throw SonataError("'node_id' must be numeric or a list of numbers");
+                }
+
+                std::vector<std::string> values;
+                for (auto& inner_el : array.items()) {
+                    values.emplace_back(inner_el.value().get<std::string>());
+                }
+
+                if (attribute == "population") {
+                    ret.emplace_back(new NodeSetBasicPopulation(std::move(values)));
+                } else {
+                    ret.emplace_back(
+                        new NodeSetBasicRule<std::string>(attribute, std::move(values)));
+                }
+            } else {
+                throw SonataError("Unknown array type");
+            }
+        }
+    }
+    return ret;
+}
+
+void parse_basic(const json& j, std::map<std::string, NodeSetRules>& node_sets) {
+    for (const auto& el : j.items()) {
+        if (el.value().is_object()) {
+            node_sets[el.key()] = _dispatch_node(el.value());
+        }
+    }
+}
+
+void check_compound(const std::map<std::string, NodeSetRules>& node_sets,
+                    const std::map<std::string, CompoundTargets>& compound_rules,
+                    size_t depth) {
+    if (depth > MAX_COMPOUND_RECURSION) {
+        throw SonataError("Compound node_set recursion depth exceeded");
+    }
+
+    for (auto const& it : compound_rules) {
+        for (auto const& name : it.second) {
+            if (compound_rules.count(name) > 0) {
+                check_compound(node_sets, compound_rules, depth + 1);
```

Comment:
I feel like I'm missing something here. How I read this, the recursion is called with the same parameters as before, but depth increased by one. To me, this would trigger as soon as any compound set refers to another one. See my test suggestion above.

Shouldn't we linearly iterate over the compound sets, and try to recursively resolve each one until max depth reached or loop detected?

Reply:
Nope, you're 100% right; great eyes.  Fixed and added test case.

> until max depth reached or loop detected?

I'm doing the simple max depth reached for now; I expect this case to be rare, so only want to throw a warning without being too specific.

Reply:
Thank you, new code looks good to me. Limits can be bumped if needed, and loop detection indeed seems like overkill here.